docs/swift/DesignDoc.md

```diff
@@ -0,0 +1,674 @@
+# Protocol Buffers in Swift
+
+## Objective
+
+This document describes the user-facing API and internal implementation of
+proto2 and proto3 messages in Apple’s Swift programming language.
+
+One of the key goals of protobufs is to provide idiomatic APIs for each
+language. In that vein, **interoperability with Objective-C is a non-goal of
+this proposal.** Protobuf users who need to pass messages between Objective-C
+and Swift code in the same application should use the existing Objective-C proto
+library. The goal of the effort described here is to provide an API for protobuf
+messages that uses features specific to Swift—optional types, algebraic
+enumerated types, value types, and so forth—in a natural way that will delight,
+rather than surprise, users of the language.
+
+## Naming
+
+*   By convention, both typical protobuf message names and Swift structs/classes
+    are `UpperCamelCase`, so for most messages, the name of a message can be the
+    same as the name of its generated type. (However, see the discussion below
+    about prefixes under [Packages](#packages).)
+
+*   Enum cases in protobufs typically are `UPPERCASE_WITH_UNDERSCORES`, whereas
+    in Swift they are `lowerCamelCase` (as of the Swift 3 API design
+    guidelines). We will transform the names to match Swift convention, using
+    a whitelist similar to the Objective-C compiler plugin to handle commonly
+    used acronyms.
+
+*   Typical fields in proto messages are `lowercase_with_underscores`, while in
+    Swift they are `lowerCamelCase`. We will transform the names to match
+    Swift convention by removing the underscores and uppercasing the subsequent
+    letter.
```

Comment:
Since you mentioned "Names with leading or doubled underscores", does Swift reserve any special meaning for names with those? I was unable to find anything in the documentation or the ABI. Or is your concern just that we should transform those as well to look more Swifty?

The single underscore appears to be treated as a keyword in Swift, in the sense that `` `_` `` is treated as an identifier named `_`, so we could escape it the way we're escaping other reserved words. Whether that's more or less desirable than a different name transformation is hard to say.

If you look at the Objective-C implementation, for example, we try to avoid just appending underscores to names, instead using suffixes like `_p` to reduce the chance of collisions. Making the generated names look good in all cases is a delicate balancing act (and not always achievable).